### PR TITLE
Update BLAS2 GER operator

### PR DESCRIPTION
**Pull request by pgorlani**
This patch introduces a new implementation of the BLAS2 GER operator.

### CODE REVIEW

File: src/operations/blas2/ger.hpp [hunk 1]
```diff
@@ -34,6 +34,173 @@
...
+  const index_t col_per_workitem = nRowsWG_ * nColsWG_ / group_size;
+  const index_t subgroup_col_id = group_local_id / nRowsWG_;
+
+  const index_t id_col0 = subgroup_col_id * col_per_workitem;
+  const index_t id_col1 = frs_col + id_col0;
+
+  value_t prefetch_lhs_ =
+      (id_row1 < dimR && id_col1 < dimC) ? lhs_.eval(id_row1, id_col1) : 0;
+
+  ndItem.barrier(cl::sycl::access::fence_space::local_space);
```
Comment (hjabird) [inline src/operations/blas2/ger.hpp:175]: I don't think `ndItem.barrier` is SYCL-2020. I think `group_barrier` should be used instead.

File: src/interface/blas2_interface.hpp [hunk 2]
```diff
@@ -1280,10 +1295,30 @@ typename sb_handle_t::event_t inline _ger(
...
+  localSize = 256;
+  useLocalMem = (_N < 8192 && _M < 8192) ? false : true;
+  nRowsWG = 32;
+  nColsWG = 32;
+#elif defined(AMD_GPU)
+  localSize = (_N < 8192 && _M < 8192) ? 512 : 256;
+  useLocalMem = (_N < 8192 && _M < 8192) ? false : true;
+  nRowsWG = (_N < 8192 && _M < 8192) ? 64 : 128;
+  nColsWG = (_N < 8192 && _M < 8192) ? 64 : 256;
+#endif
```
Comment (hjabird) [inline src/interface/blas2_interface.hpp:1318]: Is this kind of tuning normal in portBLAS?

**Review by hjabird — COMMENTED** [review]
Its been a long time since I've looked at portBLAS code, so please humour me with my questions!

File: src/interface/blas2_interface.hpp (hunk 2 above)
Comment (pgorlani) [reply src/interface/blas2_interface.hpp:1318]: So far, the performance tuning for specific device in portBLAS involved the setting template parameters.

This is not the case of this routine, in which the tuning is obtained by setting the kernel launching parameters in order to reflect specific workgroup configurations.

This is the first time it is done. If you read the removed TODO comment, this seems the way that was originally conceived for applying this kind of configuration that doesn't involve template parameters.

File: src/operations/blas2/ger.hpp [hunk 3]
```diff
@@ -34,6 +34,173 @@
...
+
+template <typename lhs_t, typename rhs_1_t, typename rhs_2_t>
+PORTBLAS_INLINE typename Ger<lhs_t, rhs_1_t, rhs_2_t>::value_t
+Ger<lhs_t, rhs_1_t, rhs_2_t>::eval(cl::sycl::nd_item<1> ndItem) {
+  using index_t = typename Ger<lhs_t, rhs_1_t, rhs_2_t>::index_t;
+
+  const index_t subgroup_size = ndItem.get_sub_group().get_local_range().get(0);
+  const index_t subgroups_per_col = nRowsWG_ / subgroup_size;
+  const index_t subgroups_per_group =
+      ndItem.get_sub_group().get_group_range().get(0);
```
[the remark this answers is not in the document]
Comment (pgorlani) [reply src/operations/blas2/ger.hpp:69]: Even if it should not be our case, I used `get(0)` for avoiding completely the possibility of unnecessary computations needed by the linear range.

File: src/operations/blas2/ger.hpp (hunk 1 above)
Comment (pgorlani) [reply src/operations/blas2/ger.hpp:175]: Yes, you are right `ndItem.barrier` is SYCL 1.2.1. However, I would prefer to keep it for consistency with the rest of portBLAS.

File: include/operations/blas2_trees.h [hunk 4]
```diff
@@ -502,6 +502,64 @@ make_trsv(vector_t &lhs_, matrix_t &matrix_, sync_t &sync_) {
...
+ * @struct Ger
+ * @brief Tree node representing the sum of scalar-vector-vector product with a
+ * matrix, i.e., it computes lhs_ such that
+ *
+ *                lhs_ =  scalar_ * ( rhs_1_ * rhs_2_^t ) + lhs_
+ *
+ * @param lhs_ input/output matrix
+ * @param scalar_ value for scaling vector product
+ * @param rhs_1_ first input vector
+ * @param rhs_2_ second input vector
```
[the remark this answers is not in the document]
Comment (pgorlani) [reply include/operations/blas2_trees.h:515]: The oneMKL documentation you are linking has its portBLAS equivalent in `include/interface/blas2_interface.h`. I cleaned it up in  https://github.com/codeplaysoftware/portBLAS/pull/505/commits/e42605a45d0608d901f51e8a5abb39c2b8060960. 

The documentation you are referring belongs to the *leaf* representing the GER kernel, `lhs_` is a matrix view which already embeds the size information, so there is no need of mentioning the size.

**Review by hjabird — APPROVED** [review]

**Pull request merged**